### DISCUSSION TITLE
How handle text fragment in pdf?

### DISCUSSION THREAD

**Discussion by Muhammadraafat1** [pymupdf/PyMuPDF#3547] [Looking for help] [answered]
My code works well (let's say in pdf1), but when I applied it to (pdf2) it didn't work well specific in redaction address.
for example we should replace the this (word="610 E Griffin Pkwy Suite D Mission, TX", replacement="310 W Palmirs Blvd Suite F Antonio, NM")
the output of pdf1:
![image](https://github.com/pymupdf/PyMuPDF/assets/140750834/e224e8dc-fb42-42f4-9e21-f367dd51f65a)

the output of pdf2:
![image](https://github.com/pymupdf/PyMuPDF/assets/140750834/39ab3e3e-9312-4da1-87be-03d7a8d6008f)

when I tried to see the difference why it works in pdf1 only, I print page.text() 
I found in the pdf2:
Address 
1: 
610 
E 
Griffin 
Pkwy 
Suite D 
Mission, 
TX 
78572

But in pdf1:
Address 1: 
610 E Griffin Pkwy Suite D Mission, TX 
78572
So, Is this the problem or not? and if it how deal with pdfs have these format of text?

**Comment by JorjMcKie** [comment, marked as answer]
I think I tried to explain this to you in another `Discussions` post already:
You obviously are dealing with OCR'ed pages. So you are not looking at actual text, but at **_images!_**
When you search / extract text, then you will get the information that your OCR engine was capable to detect. 

**_This is always error-prone!_**

The text rectangles may not exactly match the corresponding image-text (because of whatever reasons), dirt or skewed scanning may have confused the logic. Same is true for drawings: the OCR engine may think this is some text, or otherwise, your redaction / text insertion may destroy text borders that you actually wish to retain, etc., etc., and so on.

So depending on the specific situation on a page, OCR may deliver one line in one case or multiple lines / words in another case, where we as humans immediately understand that the intention is the same in both cases.

You simply have to develop code that can cope with these problems.
